File: ui/src/lib/api/update.ts
import { apiClient } from "./client";

export interface UpdateStatus {
  upgradable_count: number;
  security_count: number;
  held_count: number;
  held_packages: string[];
  last_check: string | null;
}

export interface UpgradablePackage {
  name: string;
  suite: string;
  new_version: string;
  arch: string;
  old_version: string;
  is_security: boolean;
}

export interface AptSource {
  filename: string;
  path: string;
  content: string;
  deletable: boolean;
}

export interface CommandResult {
  success: boolean;
  output: string;
  returncode: number;
}

export const updateApi = {
  status: () => apiClient.get<UpdateStatus>("/update/status").then(r => r.data),
  check: () => apiClient.post<CommandResult>("/update/check").then(r => r.data),
  upgradable: () => apiClient.get<UpgradablePackage[]>("/update/upgradable").then(r => r.data),
  upgrade: (full = false) =>
    apiClient.post<CommandResult>(`/update/upgrade?full=${full}`).then(r => r.data),
  sources: () => apiClient.get<AptSource[]>("/update/sources").then(r => r.data),
  saveSource: (filename: string, content: string) =>
    apiClient.put(`/update/sources/${filename}`, { content }).then(r => r.data),
  deleteSource: (filename: string) =>
    apiClient.delete(`/update/sources/${filename}`).then(r => r.data),
  history: (lines = 100) =>
    apiClient.get<{ log: string }>(`/update/history?lines=${lines}`).then(r => r.data),
  dpkgLog: (lines = 100) =>
    apiClient.get<{ log: string }>(`/update/dpkg-log?lines=${lines}`).then(r => r.data),
};

File: api/src/routers/update.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from src.auth.rbac import RequireAdmin, RequireViewer
from src.services import update_service

router = APIRouter(prefix="/api/update", tags=["update"])


@router.get("/status")
async def update_status(_=RequireViewer):
    return await update_service.get_status()


@router.post("/check")
async def check_updates(_=RequireAdmin):
    return await update_service.check_updates()


@router.get("/upgradable")
async def list_upgradable(_=RequireViewer):
    return await update_service.list_upgradable()


@router.post("/upgrade")
async def run_upgrade(full: bool = False, _=RequireAdmin):
    return await update_service.run_upgrade(full=full)


@router.get("/sources")
async def list_sources(_=RequireViewer):
    return await update_service.get_sources()


class SourceSave(BaseModel):
    content: str


@router.put("/sources/{filename}")
async def save_source(filename: str, body: SourceSave, _=RequireAdmin):
    try:
        return await update_service.save_source(filename, body.content)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/sources/{filename}", status_code=204)
async def delete_source(filename: str, _=RequireAdmin):
    try:
        await update_service.delete_source(filename)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/history")
async def apt_history(lines: int = 100, _=RequireViewer):
    return {"log": await update_service.get_apt_history(lines)}


@router.get("/dpkg-log")
async def dpkg_log(lines: int = 100, _=RequireViewer):
    return {"log": await update_service.get_dpkg_log(lines)}
